app/src/main/java/com/example/dogbook/main/models/Post.java

```diff
@@ -145,4 +169,73 @@ public String getTitle() {
     public String getSnippet() {
         return this.getDescription();
     }
+
+    public void like(Context context, PostReactionCallback postReactionCallback) {
+        this.likedLocally();
+        postReactionCallback.onOptimisticUpdate(this);
+
+        //Apply changes on database
+        Like like = new Like();
+        like.setAuthor(ParseUser.getCurrentUser());
+        like.setPost(this);
+        like.saveInBackground(new SaveCallback() {
+            @Override
+            public void done(ParseException e) {
+                if (e == null) {
+                    Log.i(TAG, "Like saved successfully");
+                    postReactionCallback.onSuccess(Post.this);
+                    return;
+                }
+                Log.e(TAG, "Issue saving Like", e);
+                Toast.makeText(context, "Liking was not possible", Toast.LENGTH_SHORT).show();
+                Post.this.dislikedLocally();
+                postReactionCallback.onFailure(Post.this);
+            }
+        });
+
+    }
+
+    public void dislike(Context context, PostReactionCallback postReactionCallback) {
+        this.dislikedLocally();
+        postReactionCallback.onOptimisticUpdate(this);
+        //Apply changes on database
+        ParseQuery<Like> query = ParseApplication.getLikeFromPostByUser(ParseUser.getCurrentUser(), this);
+        query.getFirstInBackground(new GetCallback<Like>() {
+            @Override
+            public void done(Like object, ParseException e) {
+                if (e == null) {
+                    object.deleteInBackground(new DeleteCallback() {
+                        @Override
+                        public void done(ParseException e) {
+                            if (e == null) {
+                                Log.i(TAG, "Dislike saved successfully");
+                                postReactionCallback.onSuccess(Post.this);
+                                return;
+
+                            }
+                            //Like was not deleted
+                            Log.e(TAG, "Issue saving dislike", e);
+                            Toast.makeText(context, "Disliking was not possible", Toast.LENGTH_SHORT).show();
+                            Post.this.likedLocally();
+                            postReactionCallback.onFailure(Post.this);
+                        }
+                    });
+                } else {
+                    //Like was not found
+                    Log.e(TAG, "Issue saving dislike", e);
+                    Toast.makeText(context, "Disliking was not possible", Toast.LENGTH_SHORT).show();
+                    Post.this.likedLocally();
+                    postReactionCallback.onFailure(Post.this);
+                }
+            }
```

Comment:
let's early return here if we go through the if block

Reply:
Wouldn't this be a problem since I make an async call inside the if? So maybe the code outside the if clause would run before hitting the return statement inside the if clause?